pkg/commands/waiter_test.go

```diff
@@ -99,18 +107,56 @@ func testWaiter(t *testing.T, when spec.G, it spec.S) {
 				Status: conditionReady(corev1.ConditionFalse, generation-1),
 			}
 
+			builderObj := &v1alpha2.Builder{
+				TypeMeta:   resourceToWatch.TypeMeta,
+				ObjectMeta: resourceToWatch.ObjectMeta,
+				Status:     v1alpha2.BuilderStatus{Status: conditionReady(corev1.ConditionTrue, generation)},
+			}
+
+			content, err := runtime.DefaultUnstructuredConverter.ToUnstructured(builderObj)
+			if err != nil {
+				panic(err)
+			}
 			watcher.addEvent(watch.Event{
-				Type: watch.Modified,
-				Object: &v1alpha2.Builder{
-					TypeMeta:   resourceToWatch.TypeMeta,
-					ObjectMeta: resourceToWatch.ObjectMeta,
-					Status:     v1alpha2.BuilderStatus{Status: conditionReady(corev1.ConditionTrue, generation)},
-				},
+				Type:   watch.Modified,
+				Object: &unstructured.Unstructured{Object: content},
 			})
 
 			require.NoError(t, waiter.Wait(context.Background(), resourceToWatch, fakeConditionChecker.conditionCheck))
 			require.True(t, fakeConditionChecker.called)
 		})
+
+		it("recovers from too old resource version error", func() {
+			watcher.addEvent(watch.Event{
+				Type: watch.Error,
+				Object: &v1.Status{
+					TypeMeta: v1.TypeMeta{
+						APIVersion: "v1",
+						Kind:       "Status",
+					},
+					Status:  "Failure",
+					Message: "too old resource version: 23358 (23360)",
+					Reason:  "Expired",
+					Code:    410,
+				},
+			})
+
+			builderObj := &v1alpha2.Builder{
+				TypeMeta:   resourceToWatch.TypeMeta,
+				ObjectMeta: resourceToWatch.ObjectMeta,
+				Status:     v1alpha2.BuilderStatus{Status: conditionReady(corev1.ConditionTrue, generation)},
+			}
+			content, err := runtime.DefaultUnstructuredConverter.ToUnstructured(builderObj)
+			if err != nil {
+				panic(err)
+			}
+			watcher.addEvent(watch.Event{
+				Type:   watch.Modified,
+				Object: &unstructured.Unstructured{Object: content},
+			})
+
+			require.NoError(t, waiter.Wait(context.Background(), resourceToWatch))
+		})
```

Comment:
This test will fail with the cachetools.Until version with the error message that was first reported. Couldn't force the too old resource version error while testing locally in my cluster :(